### ISSUE TITLE
GCM pipeline: scale up to MN

### ISSUE THREAD

**Issue by lindsayplatt** [USGS-R/lake-temperature-model-prep#258]
Considerations for scaling up to MN (and beyond 🚀 👩‍🚀)

- How long does the current GCM pipeline take since we are setup to use all lake centroids? This likely means that _most_ of the downscaled GCM cells will have at least one lake in them. Does it make since to continue with our current pattern or should we try to download the GDP results as NetCDF files and setup the pipeline to operate using those? Some thoughts on this from @jread-usgs in [this comment](https://github.com/USGS-R/lake-temperature-model-prep/issues/252#issuecomment-992542631) and in [this one](https://github.com/USGS-R/lake-temperature-model-prep/issues/252#issuecomment-993672205). Ultimately, we should probably limit the lake centroids to ones that are "model-able" (define as having some depth values?) to limit the downscaled GCM data we are pulling.
- If we do continue with the current download approach, explore and iterate on the tile size we should use to optimize download times and memory usage.
- Currently (or very soon with an upcoming PR), our data transformations to get GLM variables from GCM ones includes a couple of pivoting steps and then a series of `tidyverse` commands to `mutate`, `select`, etc. This (especially pivoting) could be very slow. Consider switching to matrix operations to increase the speed.

**Comment by lindsayplatt** [comment]
Pivoting: consider moving to `data.table::dcast()` and `data.table::melt()` instead of `tidyr::pivot_*` as ways to optimize the code quickly. Just need to convert the tibble to a data.table. Consider the following pattern for going from long to wide:

```r
tibble() %>% setDT() %>% dcast(...) %>% as_tibble()
```

**Comment by lindsayplatt** [comment]
Scaled up the time period to be the full time period for each of the 3 (1980-1999, 2040-2060, and 2080-2099) + added in all of the GCMs. Still only doing this for 5 lakes (and only 4 different cells). Download times are not looking great (16 hours in total ranging from 13 to 40 min), but the munge steps (even with all the pivoting) are taking just over 2 minutes in total.

```r
library(targets)
library(dplyr)

summarize_target_build <- function(branch_names) {
  target_info <- tar_meta(all_of(branch_names)) %>% 
    unnest(path) %>% 
    mutate(size_mb = bytes*0.000001,
           duration_min = seconds/60) %>% 
    select(path, time, duration_min, size_mb)
  
  ## PRINT OUT SOME RESULTS
  
  dur_min <- round(min(target_info$duration_min), digits = 3)
  dur_max <- round(max(target_info$duration_min), digits = 3)
  dur_tot <- round(sum(target_info$duration_min), digits = 3)
  
  message("-- DURATION --")
  message(sprintf("Shortest: %s min\nLongest: %s min\nTotal: %s min\n",
                  dur_min, dur_max, dur_tot))
          
  size_min <- round(min(target_info$size_mb), digits = 3)
  size_max <- round(max(target_info$size_mb), digits = 3)
  size_tot <- round(sum(target_info$size_mb), digits = 3)
  
  message("-- FILE SIZES --")
  message(sprintf("Smallest: %s mb\nBiggest: %s mb\nTotal: %s mb\n",
                  size_min, size_max, size_tot))
  
  # Return the full tibble of metadata
  return(target_info)
}

# How long did the GDP download take and how big are the resulting files?
tar_load(gcm_data_raw_feather)
target_metadata_download <- summarize_target_build(names(gcm_data_raw_feather))

# How long did the munge take and how big are the resulting files?
tar_load(gcm_data_daily_feather)
target_metadata_munge <- summarize_target_build(names(gcm_data_daily_feather))

```

Results:
```r
# DOWNLOADING #

-- DURATION --
Shortest: 13.427 min
Longest: 42.525 min
Total: 999.65 min

-- FILE SIZES --
Smallest: 20.735 mb
Biggest: 27.038 mb
Total: 1292.99 mb

# MUNGING #

-- DURATION --
Shortest: 0.022 min
Longest: 0.099 min
Total: 2.276 min

-- FILE SIZES --
Smallest: 0.284 mb
Biggest: 0.637 mb
Total: 22.894 mb
```

**Comment by lindsayplatt** [comment]
After converting the `pivot_longer()` call to use `data.table::melt()`:

```r
-- DURATION --
Shortest: 0.022 min
Longest: 0.078 min
Total: 1.863 min

-- FILE SIZES --
Smallest: 0.284 mb
Biggest: 0.637 mb
Total: 22.894 mb
```

After also converting the `pivot_wider()` call to use `data.table::dcast()`, interestingly resulted in a slightly longer time. Might be that more trials are needed.

```r
-- DURATION --
Shortest: 0.02 min
Longest: 0.115 min
Total: 2.051 min

-- FILE SIZES --
Smallest: 0.284 mb
Biggest: 0.637 mb
Total: 22.894 mb
```

For now, just going to share the code that would replace the `pivot_longer()` through `pivot_wider()` part of `munge_to_glm()`:

```r
# Pivot to long format first to get cells as a column.
    setDT() %>%
    melt(measure.vars = patterns("^[0-9]+$"),
                     variable.name = "cell",
                     value.name = "val",
                     variable.factor = FALSE) %>%
    .[, cell:=as.numeric(cell)] %>%

    # Now pivot wider to makes each variable a column for easy, readable munging
    # Also, removes `units` and `statistic` columns which are specific to each variable 
    # Note that data.table alphabetizes the new columns (which is probably better)
    dcast(DateTime + cell ~ variable, value.var = "val") %>%
    setDF()
```

**Comment by lindsayplatt** [comment]
I was curious if there was any kind of pattern within the download times. For just these three tiles, the 2080-2099 projection period is faster except for MIROC and MRI. This could all just be coincidence based on when my computer was doing something else, though.

```r
library(targets)
library(tidyverse)

tar_load(gcm_data_raw_feather)
branch_names <- names(gcm_data_raw_feather)
target_info <- tar_meta(all_of(branch_names)) %>% 
  unnest(path) %>% 
  mutate(size_mb = bytes*0.000001,
         duration_min = seconds/60) %>% 
  select(path, time, duration_min, size_mb) %>% 
  mutate(x = gsub("7_GCM_|_raw.feather", "", basename(path))) %>% 
  tidyr::separate(x, into = c("GCM", "start", "end", "tile"), sep= "_") %>% 
  unite("projection_period", start, end, sep = "_")

ggplot(target_info, aes(y = duration_min, x = projection_period, color = tile)) + 
  geom_point() + facet_grid(.~GCM)
```

![image](https://user-images.githubusercontent.com/13220910/148289705-d83d9e06-e636-4b42-bf7d-82055ba93dcc.png)

Same data different way to visualize/group
```r
ggplot(target_info, aes(y = duration_min, x = GCM, color = tile)) + 
  geom_point(size=4) + facet_grid(.~projection_period)   
```

![image](https://user-images.githubusercontent.com/13220910/148289810-e94adb14-2f6e-490c-84aa-268717763b92.png)

**Comment by lindsayplatt** [comment]
Current approach for testing cell & tile size:

```r
tar_load(grid_cells_sf)

# Tests:
# With current grid cell size:
#   One cell per query
#   Five cells per query
#   Ten cells per query
#   Twenty cells per query
# With smaller grid cell size:
#   One cell per query
#   Five cells per query
#   Ten cells per query
#   Twenty cells per query

# Using smaller test period and compare between this shorter period.

library(tidyverse)
library(sf)
library(geoknife)
library(retry)

run_test <- function(cell_size = 25000, n_cells = 1) {
  source("7_drivers_munge/src/GCM_driver_utils.R")
  
  default_cell_size <- 25000
  default_grid_size <- 10*default_cell_size
  
  grid_params <- tibble(
    crs = "+proj=lcc +lat_0=45 + lon_0=-97 +lat_1=36 +lat_2=52 +x_0=0 +y_0=0 +ellps=WGS84 +units=m",
    cellsize = cell_size,
    xmin = -200000,
    ymin = -1125000,
    nx = round(default_grid_size/cell_size, 0),
    ny = round(default_grid_size/cell_size, 0)
  )
  grid_cells_sf2 <- reconstruct_gcm_grid(grid_params)
  grid_cell_centroids_sf <- sf::st_centroid(grid_cells_sf) %>% mutate(tile_no = 1)
  
  set.seed(18)
  query_cells <- sample(1:100, n_cells)
  query_cell_centroids_sf <- grid_cell_centroids_sf %>%
    dplyr::filter(cell_no %in% query_cells)
  
  plot(st_geometry(grid_cells_sf))
  plot(st_geometry(grid_cells_sf2), add=TRUE, col = "red")
  plot(st_geometry(st_centroid(grid_cells_sf)), add=TRUE)
  plot(st_geometry(query_cell_centroids_sf), add=TRUE, col = "blue", cex = 5, pch = 16)
  
  query_stats <- system.time(
    out_file <- download_gcm_data(
      out_file_template = "7_drivers_munge/test/7_GCM_%s_%s_tile%s_raw.feather",
      query_geom = query_cell_centroids_sf,
      gcm_name = "GFDL",
      gcm_projection_period = "2040_2059",
      query_vars = c("pr", "tas", "qas", "rsns", "uas", "vas"),
      query_dates = c('2040-01-01 00:00:00', '2040-01-10 00:00:00')
    )
  )
  dur_sec <- query_stats[["elapsed"]]
  
  return(tibble(cell_size, n_cells, dur_sec))
}

test_criteria <- expand.grid(n_cells = c(1,5,10,20), cell_size = c(25000, 12500, 5000))

test_results <- pmap_dfr(test_criteria, run_test)
```

**Comment by lindsayplatt** [comment]
I did 3 different tests for each:

```r
n_tests <- 3
test_criteria_all <- replicate(n_tests, test_criteria, simplify = FALSE) %>% 
  bind_rows() %>% mutate(id = row_number())
test_results <- pmap_dfr(test_criteria_all, run_test)

ggplot(test_results, aes(cell_size, dur_sec)) + geom_point(size = 2) + facet_grid(.~n_cells)

ggplot(test_results, aes(n_cells, dur_sec)) + geom_point(size = 2) + facet_grid(.~cell_size)

```

![image](https://user-images.githubusercontent.com/13220910/148459339-d452dd32-e05d-4bb5-afc3-bb044422afa6.png)

![image](https://user-images.githubusercontent.com/13220910/148459449-03751350-92f5-4ba1-b0fb-818ff880465c.png)

Seeing this kind of points to the default size that we are already using `25000` being slightly better overall, but there isn't too much of a pattern that jumps out.

**Comment by lindsayplatt** [comment]
Ultimately, after this and #273 were implemented in #276, it took 16 hours to download all time for all GCMs for all of MN. See detailed results of the MN download in [this comment](https://github.com/USGS-R/lake-temperature-model-prep/pull/276#issuecomment-1013264491).
